### PR TITLE
add role-based credentials to COPY and UNLOAD command

### LINKED ISSUE #87: Support role based credentials for Copy and Unload command
We use role based credentials (http://docs.aws.amazon.com/redshift/latest/dg/copy-usage_notes-access-permissions.html#copy-usage_notes-access-role-based) instead of access and secret key. However it appears that UnloadFromSelect and CopyCommand do not support them.

### PR DESCRIPTION
**Pull request by mtrbean**
Closes #87

## Todos
- [x] MIT compatible
- [x] Tests
- [x] Documentation
- [x] Updated CHANGES.rst

### CODE REVIEW

File: tests/test_copy_command.py [hunk 1]
```diff
@@ -51,6 +52,22 @@ def test_basic_copy_case():
     assert clean(expected_result) == clean(compile_query(copy))
 
 
+def test_iam_role():
+    """Tests the use of iam role instead of access keys."""
+
```
Comment (graingert) [inline tests/test_copy_command.py]: I'd probably put iam_role as a variable here.

**Comment by graingert** [comment]
@mtrbean can you clean up your commits, and tick all the boxes as appropriate

**Comment by mtrbean** [comment]
I have moved `iam_role` to the test functions and squashed the commits.

File: sqlalchemy_redshift/commands.py [hunk 2]
```diff
@@ -18,10 +18,21 @@
...
+
+    if iam_role is not None:
+        if not IAM_ROLE_RE.match(iam_role):
+            raise ValueError(
+                'invalid IAM role ARN; does not match {pattern}'.format(
+                    pattern=IAM_ROLE_RE.pattern,
+                )
+            )
+        else:
+            return 'aws_iam_role=' + iam_role
```
Comment (jklukas) [inline sqlalchemy_redshift/commands.py]: The way this is written, `iam_role` takes precedence over `access_key_id`/`secret_access_key`/`session_token`, but the documentation doesn't make this explicit. Providing both types of credentials seems like it indicates confusion on the part of the user, and I'd propose that we raise an error in that case indicating that the two credential styles are mutually exclusive and should not be mixed.

So, if within this `if` block, we'd check if any of access key ID, secret access key, or session token are not `None`, and then raise an error.

Comment (graingert) [inline sqlalchemy_redshift/commands.py]: probably best to create two classes, so you can use the API like:

```.copy(access=APIKeyCredential(access_key_id, secret_access_key, session_token), ...)```
```.copy(access=IAMRoleCredential(iam_role), ...)```

class APIKeyCredential(object):

Comment (jklukas) [inline sqlalchemy_redshift/commands.py]: > probably best to create two classes

I'd be fine with that. Would we want to keep the existing `access_key_id, secret_access_key, session_token` options for backward-compatibility and mark them as deprecated?

Comment (graingert) [inline sqlalchemy_redshift/commands.py]: I'd like to stick with the code as is for now, maybe add a TypeError

I've added it to my personal v1-breaking-change-wishlist

Comment (graingert) [inline sqlalchemy_redshift/commands.py]: @mtrbean can you check split the iam_role into aws_account_id and iam_role_name validating both against a regex.

```python
if iam_role_name is not None:
    if not AWS_ACCOUNT_ID_RE.match(aws_account_id):
        raise ValueError(...)
    if not IAM_ROLE_NAME_RE.match(iam_role_Name):
        raise ValueError(...)

    credentials = 'aws_iam_role=arn:aws:iam::{aws_acount_id}:role/{iam_role_name}'.format(...)
```

Comment (jklukas) [inline sqlalchemy_redshift/commands.py]: > I'd like to stick with the code as is for now, maybe add a TypeError

I'm fine with that.

Comment (graingert) [inline sqlalchemy_redshift/commands.py]: currently the iam_role_name is .* allowing "; DROP TABLE foobar; --"

Comment (graingert) [inline sqlalchemy_redshift/commands.py]: "Maximum 64 characters. Use alphanumeric and '+=,.@-_' characters"

**Comment by mtrbean** [comment]
Changes made, let me know if anything further changes needed

**Comment by graingert** [comment]
looks good, squash your commits though

**Comment by mtrbean** [comment]
Squashed. I think you can also make Github squash the commits for you when you merge.

**Comment by graingert** [comment]
I can it's nicer not to though as it looses the merge commit

**Comment by mtrbean** [comment]
@graingert can this be merged?

**Comment by graingert** [comment]
@mtrbean merged! thanks :)

**Pull request merged**